rpcs3/Emu/Cell/SPUThread.cpp

```diff
@@ -4486,28 +4510,33 @@ bool spu_thread::stop_and_signal(u32 code)
 
 		spu_log.trace("sys_spu_thread_tryreceive_event(spuq=0x%x)", spuq);
 
-		std::lock_guard lock(group->mutex);
+		std::shared_ptr<lv2_event_queue> queue;
 
-		lv2_event_queue* queue = nullptr;
+		reader_lock{group->mutex}, queue = get_queue(spuq);
 
-		for (auto& v : this->spuq)
+		std::unique_lock<shared_mutex> qlock, group_lock;
+
+		while (true)
 		{
-			if (spuq == v.first)
+			if (!queue)
 			{
-				if (lv2_obj::check(v.second))
-				{
-					queue = v.second.get();
-					break;
-				}
+				return ch_in_mbox.set_values(1, CELL_EINVAL), true;
 			}
-		}
 
-		if (!queue)
-		{
-			return ch_in_mbox.set_values(1, CELL_EINVAL), true;
-		}
+			qlock = std::unique_lock{queue->mutex};
+			group_lock = std::unique_lock{group->mutex};
 
-		std::lock_guard qlock(queue->mutex);
+			if (auto queue0 = get_queue(spuq); queue != queue0)
```

Comment:
Since `get_queue` returns a reference, it may be not a good idea to use `auto` here

Reply:
idk about that, I've made the intention clearer.